src/Runtime/ServiceManager.php

```diff
@@ -372,22 +372,60 @@ public function instantiateServices(array $bootstrapCommandClasses, DrushContain
         return $commandHandlers;
     }
 
-    /**
-     * Determine if the first parameter of the create method supports our container with delegate.
-     */
-    protected function supportsCompoundContainer($class, $drush_container): bool
+    protected function getStaticCreateFactoryArguments(string $class, DrushContainer $drushContainer, ?DrupalContainer $drupalContainer = null): ?array
     {
+        if (!method_exists($class, 'create')) {
+            return null;
+        }
+
         $reflection = new \ReflectionMethod($class, 'create');
-        $hint = (string)$reflection->getParameters()[0]->getType();
-        return is_a($drush_container, $hint);
+        if (!$reflection->isStatic()) {
+            return null;
+        }
+
+        $params = $reflection->getParameters();
+
+        $args = [];
+        $type = ltrim((string) $params[0]->getType(), '?');
+        if ($drupalContainer && is_a($type, DrupalContainer::class, true)) {
+            // The factory create() method explicitly expects a Drupal container as 1st argument.
+            $args[] = $drupalContainer;
+        } elseif (is_a($type, DrushContainer::class, true)) {
+            // The factory create() method explicitly expects a Drush container as 1st argument. Don't add a 2nd
+            // argument. If the Drupal container has been initialized, it's already a delegate.
+            return [$drushContainer];
+        } elseif ($drupalContainer && is_a($type, ContainerInterface::class, true)) {
+            // The factory create() method expects a container of any type as st argument and the Drupal container has
+            // been initialized. Pass it as 1st argument.
+            $args[] = $drupalContainer;
+        }
+
+        // Add Drush container as 2nd argument if the method expects one.
+        if (isset($params[1])) {
+            $type = ltrim((string) $params[1]->getType(), '?');
+            if (is_a($type, ContainerInterface::class, true)) {
+                $args[] = $drushContainer;
+            }
+        }
```

Comment:
I dont like commands and AutowireTrait having multiple numbers of params and type hints in create(). How about if we ask commands going forward to always type hint to `Psr\Container\ContainerInterface`. That receives the compound container. AutowireTrait could pull the drupalContainer out and get container params as needed. We need to declare a service for the drupalContainer in the league container. [We already store a reference to the league container inside itself](https://github.com/drush-ops/drush/blob/5b1d3d8a7c518d2882d604bfcd14ee3902fc568f/src/Runtime/DependencyInjection.php#L69). So this would be a similar service.